### PR TITLE
snapshot for AWS platform

### PR DESCRIPTION
**Pull request by ropatil010**
From version 4.7+
/cc @duanwei33 @chao007 @jhou1 @Phaow

### CODE REVIEW

**Comment by ropatil010** [comment]
Upgrade prepare logs
https://privatebin-it-iso.int.open.paas.redhat.com/?510e7b4ea4ecd2ae#H36ZeJCKFeNZqubGSqeJXXJkBWiaEWzeUXiwU3fF3eYQ

Upgrade check logs
https://privatebin-it-iso.int.open.paas.redhat.com/?c2cabd01e6f39bfb#AfUdUBC8WCxhy23U7P3Y7XeiuXebtDXo3YXLGThra4DQ

File: features/upgrade/storage/upgrade.feature [hunk 1]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
...
+    #Snapshot operator/controller status checking
+    Given the "csi-snapshot-controller" operator version matches the current cluster version
+
+    Given the status of condition "Degraded" for "csi-snapshot-controller" operator is: False
+    Given the status of condition "Progressing" for "csi-snapshot-controller" operator is: False
+    Given the status of condition "Available" for "csi-snapshot-controller" operator is: True
+    Given the status of condition "Upgradeable" for "csi-snapshot-controller" operator is: True
+
+    When I run the :new_project client command with:
+      | project_name  | snapshot-upgrade-ocp-44578  |
```
Comment (jhou1) [inline features/upgrade/storage/upgrade.feature:365]: Style suggestion: make sure there is only one space between `project_name` and `|` in the table.

File: features/upgrade/storage/upgrade.feature [hunk 2]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
...
+    When I run oc create over "pod.yaml" replacing paths:
+      | ["metadata"]["name"]                                         | mypod-ori    |
+      | ["spec"]["volumes"][0]["persistentVolumeClaim"]["claimName"] | mypvc-ori    |
+      | ["spec"]["containers"][0]["volumeMounts"][0]["mountPath"]    | /mnt/csipath |
+
+    Then the step should succeed
+    Given the pod named "mypod-ori" becomes ready
+
+    When I execute on the pod:
+      | touch   | /mnt/csipath/test-before-upgrade         |
```
Comment (jhou1) [inline features/upgrade/storage/upgrade.feature:383]: Style suggestion: make sure there is only one space between `touch` and `|` in the table.  and the next table

File: features/upgrade/storage/upgrade.feature [hunk 3]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
...
+    Given the status of condition "Degraded" for "csi-snapshot-controller" operator is: False
+    Given the status of condition "Progressing" for "csi-snapshot-controller" operator is: False
+    Given the status of condition "Available" for "csi-snapshot-controller" operator is: True
+    Given the status of condition "Upgradeable" for "csi-snapshot-controller" operator is: True
+
+    When I run the :new_project client command with:
+      | project_name  | snapshot-upgrade-ocp-44578  |
+    When I use the "snapshot-upgrade-ocp-44578" project
+
+    Given I obtain test data file "storage/misc/pvc-with-storageClassName.json"
```
Comment (jhou1) [inline features/upgrade/storage/upgrade.feature:368]: Do we delete this storageclass after `@upgrade-check`?

File: features/upgrade/storage/upgrade.feature [hunk 4]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
...
+
+    Given I obtain test data file "storage/misc/pod.yaml"
+    When I run oc create over "pod.yaml" replacing paths:
+      | ["metadata"]["name"]                                         | mypod-restore |
+      | ["spec"]["volumes"][0]["persistentVolumeClaim"]["claimName"] | mypvc-restore |
+      | ["spec"]["containers"][0]["volumeMounts"][0]["mountPath"]    | /mnt/csipath  |
+    Given the pod named "mypod-restore" becomes ready
+
+    When I execute on the pod:
+      | ls | /mnt/csipath   |
```
Comment (jhou1) [inline features/upgrade/storage/upgrade.feature:425]: Similar style suggestion for table

**Comment by jhou1** [comment]
I see the test does this:

`@upgrade-prepare`
- create data
- snapshot
- restore

`@upgrade-check`
- create data and verify data consistency

Per the upgrade test scenario, I'm not sure how that's relevant to the upgrade. I think the snapshot/restore scenarios had better run during `@upgrade-check`. Any suggestions from other storage members?

**Comment by duanwei33** [comment]
Hi @jhou1, that's a good catch, I missed it in the off-line review.

Actually, I think we missed one step to create a new pvc restored from snapshotshot(which was created before the upgrade), and using the new pvc to create pod instead of using the pvc which was created before the upgrade.

I thinks the whole test looks like this:
@upgrade-prepare
* create data
* snapshot
* restore pvc
All steps above check that snapshot works before the upgrade and the snapshot will be used in @upgrade-check

@upgrade-check
* restore pvc from snapshot which created before the upgrade
* create pod and verify data consistency
* Also, we could consider having another round after upgrade: create new data, snapshot, restore pvc 
All steps above check the previous snapshot could be restored and the new snapshot could be created and restored.



> I see the test does this:
> 
> `@upgrade-prepare`
> 
> * create data
> * snapshot
> * restore
> 
> `@upgrade-check`
> 
> * create data and verify data consistency
> 
> Per the upgrade test scenario, I'm not sure how that's relevant to the upgrade. I think the snapshot/restore scenarios had better run during `@upgrade-check`. Any suggestions from other storage members?

File: features/upgrade/storage/upgrade.feature [hunk 5]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
...
+    Then the output should contain:
+      | test-before-upgrade |
+
+  # @author ropatil@redhat.com
+  @upgrade-check
+  @users=upuser1,upuser2
+  @admin
+  Scenario: Snapshot operator should work well before and after upgrade of cluster on AWS platform - check
```
Comment (duanwei33) [inline features/upgrade/storage/upgrade.feature:433]: "- check" need to be removed.

File: features/upgrade/storage/upgrade.feature (hunk 3 above)
Comment (ropatil010) [reply features/upgrade/storage/upgrade.feature:368]: Using default storage class(gp2-csi). Not creating new sc.

File: features/upgrade/storage/upgrade.feature (hunk 5 above)
Comment (ropatil010) [reply features/upgrade/storage/upgrade.feature]: Thanks @duanwei33 , Changed the steps in upgrade-check as per your suggestions.

**Comment by ropatil010** [comment]
Upgrade prepare logs:
https://privatebin-it-iso.int.open.paas.redhat.com/?915e45e0c1e5e979#aofziuSCkWgcf9o9P1Q91nd81eZs9YxMYvZeT2ppLur

Upgrade check logs: https://privatebin-it-iso.int.open.paas.redhat.com/?d95284dbd07f0723#85Ca3dWbgweNw3JwD3pzvJT1cmZUadCN33yANj7eeSs9

Pods,pvc logs:
https://privatebin-it-iso.int.open.paas.redhat.com/?7eeb014095483681#GwyuGrL2tvRCLbqmB43E3TLRrRLJdL6z2fPd5v5R19YT

File: features/upgrade/storage/upgrade.feature [hunk 6]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
...
+    Then the step should succeed
+    Given the pod named "mypod-ori" becomes ready
+
+    When I execute on the pod:
+      | touch   | /mnt/csipath/test-before-upgrade         |
+    Then the step should succeed
+    And I execute on the pod:
+      | sh | -c | sync -f /mnt/csipath/test-before-upgrade |
+
+    Given I ensure "mypod-ori" pod is deleted
```
Comment (chao007) [inline features/upgrade/storage/upgrade.feature]: No need to remove this pod "mypod-ori"

Comment (ropatil010) [reply features/upgrade/storage/upgrade.feature]: Thanks @chao007 .
I will try to change the code, As per your suggestions to replace the pod with deployment in entire snapshot test.

**Comment by ropatil010** [comment]
Hi All,
Kindly review the changes. Submitted in the 4th commit with name "snapshotwithdep"
@chao007 As per your suggestion replaced pod with deployment, Executed and verified. 

Logs:
Upgrade Prepare: 
https://privatebin-it-iso.int.open.paas.redhat.com/?d680e4c26b9d95ca#4tW4PVsctmAdXmTJXDbBGSP1qnPfzf3M1RjSX8gJvGa1

Upgrade check:
https://privatebin-it-iso.int.open.paas.redhat.com/?904c49e35a9f3d95#8jL8QdohSXeHRWNApKncbDPjceZrxVENgBysFHZpMZXJ

pvc,pod,volumesnapshot contentdetails:
https://privatebin-it-iso.int.open.paas.redhat.com/?cf5ee4dd1486cdd9#CQGintEt8ahkGPvuupALQUnmmk81m2Dmi88ybyHd1xMB

File: features/upgrade/storage/upgrade.feature [hunk 7]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
...
+  @users=upuser1,upuser2
+  @admin
+  Scenario: Snapshot operator should work well before and after upgrade of cluster on AWS platform - prepare
+    Given I switch to cluster admin pseudo user
+    Given the master version >= "4.7"
+      
+    #Snapshot operator/controller status checking
+    Given the "csi-snapshot-controller" operator version matches the current cluster version
+
+    Given the status of condition "Degraded" for "csi-snapshot-controller" operator is: False
```
Comment (Phaow) [inline features/upgrade/storage/upgrade.feature:365]: @ropatil010 @duanwei33 From my side, the status of csi-snapshot-controller cluster check just need one. Check the status of `condition “Available” for “csi-snapshot-controller” operator is True` is not enough? Could you please help to explain the doubt for me? Thanks!

File: features/upgrade/storage/upgrade.feature [hunk 8]
```diff
@@ -344,3 +344,120 @@ Feature: Storage upgrade tests
       | ls | /mnt/hostpath |
     Then the output should contain:
       | test-before-upgrade |
+
+  # @author ropatil@redhat.com
+  @upgrade-prepare
+  @users=upuser1,upuser2
+  @admin
+  Scenario: Snapshot operator should work well before and after upgrade of cluster on AWS platform - prepare
```
Comment (Phaow) [inline features/upgrade/storage/upgrade.feature:358]: I see it in polarion it's a new case. The scenario named `"[AWS-EBS] [Snapshot operator] should work well before and after upgrade of a cluster"` is better.

File: features/upgrade/storage/upgrade.feature [hunk 9]
```diff
@@ -344,3 +344,222 @@ Feature: Storage upgrade tests
...
       | test-before-upgrade |
+
+  # @author ropatil@redhat.com
+  # Description:
+  # Check csi snapshot controller status 
+  # Create project, pvc, deployment, file(test-before-upgrade) in deployment
+  # Create Volumesnapshot Class, Volumesnapshot
+  # Create Restore pvc/dep
+  # Check file(test-before-upgrade) in dep
+  @upgrade-prepare
```
Comment (jhou1) [inline features/upgrade/storage/upgrade.feature:355]: Do you mind removing these comments from the scenario header? Comments could be added right before each step.

File: features/upgrade/storage/upgrade.feature [hunk 10]
```diff
@@ -344,3 +344,222 @@ Feature: Storage upgrade tests
...
+
+  # @author ropatil@redhat.com
+  # Description:
+  # Check csi snapshot controller status after upgrade
+  # Use project, create restore pvc/deployment, file(test-after-upgrade) in deployment
+  # Check both files(test-before-upgrade,test-after-upgrade) in dep 
+  # Create new project, new pvc/dep, file(test-after-upgrade) in dep
+  # Create new Volumesnapshot Class, Volumesnapshot
+  # Create new Restore pvc/dep,file(test-after-upgrade-check) in dep
+  # Check files(test-after-upgrade,test-after-upgrade-check) in dep
```
Comment (jhou1) [inline features/upgrade/storage/upgrade.feature:447]: Do you mind removing these comments from the scenario header? Comments could be added to right before each step.

File: features/upgrade/storage/upgrade.feature (hunk 7 above)
Comment (jhou1) [reply features/upgrade/storage/upgrade.feature:359]: These steps could make the verification more solid, making sure the cluster operator is available and does not have any unhealthy state

**Comment by jhou1** [comment]
Looks good, once you squash the commits we could merge.

**Pull request merged**